### ISSUE TITLE
Discretization of disk footprints (also ellipse, ball)

### ISSUE THREAD

**Issue by grlee77** [scikit-image/scikit-image#5470]
## Description

When constructing footprints (structuring elements) corresponding to a disk, the following code is currently used:

https://github.com/scikit-image/scikit-image/blob/d85b06468e063f2b8aa6aa7e073e120294e8e2f8/skimage/morphology/footprints.py#L118-L120

A second possibility is the following:
```Python
return np.array((X ** 2 + Y ** 2) <= (radius + 0.5) ** 2, dtype=dtype)
```
where the radius has been extended by 0.5 voxels. Practially the former means the edge of the disk falls at the voxel center a distance radius away, while for the later, the disk extends to the outer edge of that same voxel. A concrete illustration of this is given for radius = 3 (vs. 3.5) below:

![morphology_discrete_illustration_sm](https://user-images.githubusercontent.com/6528957/125207719-caf6b800-e25b-11eb-8c42-464e16de2502.png)

Below is an illustration where the top row are the currently returned footprints for `skimage.morphology.disk` and the bottom row are the footprints with the proposed modification.

![morphology_disks](https://user-images.githubusercontent.com/6528957/125206402-1c4f7900-e255-11eb-8261-6418edf3607f.png)

I don't know that the existing behavior is technically a bug, but it is more a matter of what convention is chosen. To me, the lower row look visually more disk-like (although radius=1 is too coarse to resemble a disk in either case).

If this change is made for `disk`, the same convention should be used for `ball` and `ellipse` as well.

**Comment by grlee77** [comment]
Additional considerations:
1.) `calling skimage.morphology.footprints.ellipse(radius, radius)` resembles the "proposed" disks more than the "current" ones, although they are not identical for all choices of radius.
2.) The "current" implementation is consistent with the disks produced by Matlab (there are existing tests that verify this)
3.) OpenCV has a "MORPH_ELLIPSE" shape, but apparently not a disk one. If I make ellipses of the corresponding size they actually end up slightly asymmetric, resembling "current" along one axis, but "proposed" along the other! (seems undesirable)

We could potentially allow both conventions by adding a keyword-only `strict_radius` parameter with default value of True corresponding to the current behavior and a value of False corresponding to the proposed shape here.

![morph_ocv](https://user-images.githubusercontent.com/6528957/126218165-ff0e891b-c457-41d8-b4ae-cb1c3b6a3ba5.png)

**Comment by grlee77** [comment]
ImageMagick deals with this by allowing specification of a floating point radius which for example matches "current" when the radius is a whole number or "proposed" if the radius is a half pixel number. Size is always ``(floor(radius) * 2 + 1, floor(radius) * 2 + 1)``
see: https://legacy.imagemagick.org/Usage/morphology/#disk

**Comment by rfezzani** [comment]
Hi @grlee77, and thank you for raising this. I agree, you proposed implementation looks visually more disk-like!
I like the idea of `strict_radius` argument :+1:

**Comment by joshuacwnewton** [comment]
Hi there! Just following up on this issue with a quick summary of the state of things in 2025, since we ran into issues while using the `ellipse` footprint function, but the issue lacks context for why it was reopened (which would have helped us, hehe).

#### 1. `disk` + `ball`

> I like the idea of `strict_radius` argument 👍

The `disk` and `ball` structuring elements presently have the `strict_radius` parameter, added in [#6151](https://github.com/scikit-image/scikit-image/pull/6151).

#### 2. `ellipse`

> 1.) calling `footprints.ellipse(radius, radius)` resembles the "proposed" disks more than the "current" ones, although they are not identical for all choices of radius.

I can reproduce this "not identical" behavior.

- Expected (i.e. `disk(radius=5, strict_radius=False)`) (sample: 5x5):
  <img width="1365" height="315" alt="Image" src="https://github.com/user-attachments/assets/51044dbb-8931-479b-bd32-84aa3600ba13" />
- Actual (sample: 5x5):
  <img width="1365" height="315" alt="Image" src="https://github.com/user-attachments/assets/a541f392-62ab-4737-8c2a-8c3560934aba" />
- Curiously, `ellipse` uses [`draw.ellipse`](https://github.com/scikit-image/scikit-image/blob/83a65ac0288bdc52f262b89b7f4c12c1324f1f4f/skimage/morphology/footprints.py#L713) internally, while `disk` and `ball` use equations directly in their functions. 
- `skimage` specifies a bounding box +2 larger than the actual structuring element (13x13 instead of 11x11 in this case):
  https://github.com/scikit-image/scikit-image/blob/83a65ac0288bdc52f262b89b7f4c12c1324f1f4f/skimage/morphology/footprints.py#L713
- `draw.ellipse` uses `draw._ellipse_in_shape`, which in turn uses the following equation:
  https://github.com/scikit-image/scikit-image/blob/83a65ac0288bdc52f262b89b7f4c12c1324f1f4f/skimage/draw/draw.py#L48-L50
  <img width="1615" height="368" alt="Image" src="https://github.com/user-attachments/assets/9ba37256-73ac-435a-b056-317c984a022c" />
- Then, because `< 1` is used instead of `<= 1`, we get this shape: 
  https://github.com/scikit-image/scikit-image/blob/83a65ac0288bdc52f262b89b7f4c12c1324f1f4f/skimage/draw/draw.py#L51
  <img width="1615" height="368" alt="Image" src="https://github.com/user-attachments/assets/a732903d-bfbe-4cad-b1f6-8505412add87" />
  And then the nonzero pixels are properly centered in the smaller 11x11 bounding box.

tl;dr ellipse uses an entirely different procedure that isn't quite comparable to what one would expect after using both `radius_strict` options for `disk` and `ball`.

#### 3. potential fix

For my project's use, my colleague [developed an implementation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d3329c41f4be9073d3d21c6f66f3cd855973987c/spinalcordtoolbox/math.py#L58-L76) to produce a variable-sized (incl. non-circular) ellipsoid/spheroid shape. We use the `strict_radius=True` equation by default, but I presume it could be extended to add the option `+ 0.5` for `strict_radius=False` behavior.

Given that `square` and `cube` structuring elements were recently replaced with the unified `footprint_rectangle` interface, perhaps this issue could be resolved in the same manner by coming up with a unified interface for curved structuring elements as well?

EDIT: Ah, this is in fact a current WIP:

- https://github.com/scikit-image/scikit-image/pull/7628

And @lagru already seems to be aware of the challenges involved with defining ellipses: https://github.com/scikit-image/scikit-image/pull/7628#discussion_r1943576597 (though, tagging just in case this issue can provide possible ideas to help move the PR along).

**Comment by lagru** [comment]
Thank you for reaching out @joshuacwnewton. It's reassuring that you can confirm many of the conclusions I arrived at in #7628. It's been a while and that PR has stalled a bit because of uncertainty around how to handle the transition while keeping backwards-compatiblity. Since you have some recent context, any insights would be appreciated. Especially with the annoying inconsistencies pointed out in https://github.com/scikit-image/scikit-image/pull/7628#discussion_r1943576597. 

As a downstream-project, do you have a preference for a specific algorithm?

**Comment by joshuacwnewton** [comment]
> Since you have some recent context, any insights would be appreciated. Especially with the annoying inconsistencies pointed out in https://github.com/scikit-image/scikit-image/pull/7628#discussion_r1943576597.

Sure thing! I'll talk it over with my colleague who drafted the currently-open PR in our project. :)

> As a downstream-project, do you have a preference for a specific algorithm?

I'm not sure we have a preference? Our use case involves generating dilated masks around spinal cord segmentations, and so the radii we use (15-30+ px) have been large enough that differences in implementation aren't very apparent for us, I think.

For what it's worth, we've been satisfied with the existing `disk` and `ball` implementations for years (until we ran into a case where we wanted a shape with different bounding box dimensions and found `ellipse` to not suit our needs). So, I imagine preserving the existing behavior that users have come to expect of `disk` and `ball` might be worthwhile? Not to mention that the `radius_strict` behavior feels the most mathematically intuitive to me (i.e. if I'm specifying a radius of `N`, I would be surprised to see a pixel `<N` distance away set to `1` in the footprint). Plus the consistency with MATLAB as highlighted above makes a compelling argument to me.

Subjectively, though, I do agree that the `radius_strict=False` behavior does feel more "disk-like"? But I think offering a parameter to vary the amount of padding should cover the needs of users who want a more visually pleasing disk. :)
